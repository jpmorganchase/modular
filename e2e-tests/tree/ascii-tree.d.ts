declare module 'ascii-tree' {
  export function generate(input: string): string;
}
