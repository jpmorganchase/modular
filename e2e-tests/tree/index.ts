import fs from 'fs';
import path from 'path';
import asciiTree from 'ascii-tree';
import hash from '@emotion/hash';

function times(str: string, length: number) {
  return Array.from({ length }, () => str).join('');
}

const defaultIgnores = ['node_modules', '.git', '.DS_Store'];
const defaultHashIgnores = [
  // adding lockfiles here because it can be different
  // on different runs; since we install the latest versions
  // of some packages when making a repository
  'yarn.lock',
  'package-lock.json',
  // adding package.json files since versions within these get bumped
  'package.json',
];

function tree(
  _path: string,
  level = 1,
  options = {
    ignores: defaultIgnores,
    hashIgnores: defaultHashIgnores,
  },
): string {
  const stat = fs.statSync(_path);
  if (stat.isDirectory()) {
    const children = fs.readdirSync(_path);
    const dirArr = _path.split('/');
    const dir = dirArr[dirArr.length - 1];
    // todo - should these be sorted?
    // todo - handle symlinks, etc
    return `${times('#', level)}${dir}\n${children
      .filter((child: string) => !options.ignores.includes(child))
      .map((child: string) => tree(path.join(_path, child), level + 1, options))
      .join('\n')}`;
  } else {
    return [
      `${times('#', level)}${path.basename(_path)}`,
      options.hashIgnores.includes(path.basename(_path))
        ? undefined
        : `#${hash(fs.readFileSync(_path, 'utf8'))}`,
    ]
      .filter(Boolean)
      .join(' ');
  }
}

export default function generate(_path: string): string {
  return asciiTree.generate(tree(_path));
}
