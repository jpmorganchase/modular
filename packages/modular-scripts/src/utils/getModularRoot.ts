import * as path from 'path';
import * as fs from 'fs-extra';
import findUp from 'find-up';

function isModularRoot(packageJson: { modular?: Record<string, unknown> }) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory: string) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

export default function getModularRoot(): string {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
}
