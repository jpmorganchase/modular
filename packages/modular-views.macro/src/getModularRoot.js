// keep this in sync with modular-scripts/src/getModularRoot
const path = require('path');
const fs = require('fs-extra');
const findUp = require('find-up');

function isModularRoot(packageJson) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

module.exports = function getModularRoot() {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
};
