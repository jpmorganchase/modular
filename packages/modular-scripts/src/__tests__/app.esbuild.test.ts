import execa from 'execa';
import rimraf from 'rimraf';
import tree from 'tree-view-for-tests';
import path from 'path';
import fs from 'fs-extra';
import prettier from 'prettier';

import getModularRoot from '../utils/getModularRoot';

const modularRoot = getModularRoot();

// These tests must be executed sequentially with `--runInBand`.

const packagesPath = path.join(getModularRoot(), 'packages');

function modular(str: string, opts: Record<string, unknown> = {}) {
  return execa('yarnpkg', ['modular', ...str.split(' ')], {
    cwd: modularRoot,
    cleanup: true,
    // @ts-ignore
    env: {
      USE_MODULAR_ESBUILD: 'true',
    },
    ...opts,
  });
}

function cleanup() {
  rimraf.sync(path.join(packagesPath, 'sample-esbuild-app'));
  rimraf.sync(path.join(modularRoot, 'dist/sample-esbuild-app'));

  // run yarn so yarn.lock gets reset
  return execa.sync('yarnpkg', ['--silent'], {
    cwd: modularRoot,
  });
}

beforeAll(cleanup);
afterAll(cleanup);

describe('when working with an app', () => {
  beforeAll(async () => {
    await modular('add sample-esbuild-app --unstable-type app', {
      stdio: 'inherit',
    });

    await modular('build sample-esbuild-app', {
      stdio: 'inherit',
    });
  });

  it('can add an app', () => {
    expect(tree(path.join(packagesPath, 'sample-esbuild-app')))
      .toMatchInlineSnapshot(`
      "sample-esbuild-app
      ├─ README.md #g3sy9z
      ├─ package.json
      ├─ public
      │  ├─ favicon.ico #6pu3rg
      │  ├─ index.html #1m6toxd
      │  ├─ logo192.png #1nez7vk
      │  ├─ logo512.png #1hwqvcc
      │  ├─ manifest.json #19gah8o
      │  └─ robots.txt #1sjb8b3
      ├─ src
      │  ├─ App.css #1o0zosm
      │  ├─ App.tsx #c80ven
      │  ├─ __tests__
      │  │  └─ App.test.tsx #1u72nad
      │  ├─ index.css #o7sk21
      │  ├─ index.tsx #zdn6mw
      │  ├─ logo.svg #1okqmlj
      │  └─ react-app-env.d.ts #t4ygcy
      └─ tsconfig.json #6rw46b"
    `);
  });

  it('can build an app', () => {
    expect(tree(path.join(modularRoot, 'dist', 'sample-esbuild-app')))
      .toMatchInlineSnapshot(`
      "sample-esbuild-app
      ├─ favicon.ico #6pu3rg
      ├─ index.html #zk3xji
      ├─ logo192.png #1nez7vk
      ├─ logo512.png #1hwqvcc
      ├─ manifest.json #19gah8o
      ├─ package.json
      ├─ robots.txt #1sjb8b3
      └─ static
         ├─ css
         │  ├─ index-PE2NG66F.css #1t4q6xl
         │  └─ index-PE2NG66F.css.map #za6yi0
         ├─ js
         │  ├─ index-TKIUYRMA.js #l9v7el
         │  └─ index-TKIUYRMA.js.map #51gooi
         └─ media
            └─ logo-PGX3QVVN.svg #1okqmlj"
    `);
  });

  it('can generate a index.html', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(modularRoot, 'dist', 'sample-esbuild-app', 'index.html'),
          ),
        ),
        {
          filepath: 'index.html',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a index.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'sample-esbuild-app',
              'static',
              'js',
              'index-TKIUYRMA.js',
            ),
          ),
        ),
        {
          filepath: 'index.js',
        },
      ),
    ).toMatchSnapshot();
  });

  type SourceMap = Record<string, string | string[]>;

  const readSourceMap = (pathName: string) => {
    const map = fs.readJsonSync(
      path.join(modularRoot, 'dist', 'sample-esbuild-app', pathName),
    ) as SourceMap;
    return {
      ...map,
      // make the source root //modular so that it's the same
      // across platforms
      sourceRoot: '//modular',
    };
  };

  it('can generate a index.js.map', () => {
    expect(readSourceMap('static/js/index-TKIUYRMA.js.map')).toMatchSnapshot();
  });

  it('can generate a index.css.map', () => {
    expect(
      readSourceMap('static/css/index-PE2NG66F.css.map'),
    ).toMatchSnapshot();
  });
});
