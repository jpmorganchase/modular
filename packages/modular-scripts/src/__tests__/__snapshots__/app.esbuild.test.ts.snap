// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when working with an app can generate a index.html 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\" />
    <link rel=\\"icon\\" href=\\"/favicon.ico\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"theme-color\\" content=\\"#000000\\" />
    <meta
      name=\\"description\\"
      content=\\"Web site created using create-modular-app\\"
    />
    <link rel=\\"apple-touch-icon\\" href=\\"/logo192.png\\" />
    <link rel=\\"manifest\\" href=\\"/manifest.json\\" />
    <title>React App</title>
    <link rel=\\"stylesheet\\" href=\\"/index.css\\" />
  </head>
  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id=\\"root\\"></div>
    <script type=\\"module\\" src=\\"/index.js\\"></script>
  </body>
</html>
"
`;
